pyzx/graph/base.py

```diff
@@ -825,38 +819,241 @@ def add_edge_smart(self, e: ET, edgetype: EdgeType.Type):
         """Like add_edge, but does the right thing if there is an existing edge."""
         self.add_edge_table({e : [1,0] if edgetype == EdgeType.SIMPLE else [0,1]})
 
-    def set_phase_master(self, m: 'simplify.Simplifier') -> None:
-        """Points towards an instance of the class :class:`~pyzx.simplify.Simplifier`.
-        Used for phase teleportation."""
-        self.phase_master = m
-
-    def update_phase_index(self, old:VT, new:VT) -> None:
-        """When a phase is moved from a vertex to another vertex,
-        we need to tell the phase_teleportation algorithm that this has happened.
-        This function does that. Used in some of the rules in `simplify`."""
-        if not self.track_phases: return
-        i = self.phase_index[old]
-        self.phase_index[old] = self.phase_index[new]
-        self.phase_index[new] = i
-
-    def fuse_phases(self, p1: VT, p2: VT) -> None:
-        if p1 not in self.phase_index or p2 not in self.phase_index:
+    def set_phase_teleporter(self, teleporter: 'simplify.PhaseTeleporter', fusing_mode: bool = True) -> None:
+        """Used for phase teleportation.
+        If ``fusing_mode`` is True then phases will be tracked as the graph is simplified.
+        Otherwise info about previously teleported phases from ``teleporter`` is stored, but not placed on the graph yet.
+        They will then be placed throughout simplification when required, or through the function :func:`place_tracked_phases`
+
+        :param teleporter: Instance of the class :class:`~pyzx.simplify.PhaseTeleporter`
+        :param fusing_mode: Defaults to True
+        """
+        self.phase_tracking = True
+        
+        if fusing_mode:
+            self.phase_teleporter = teleporter
             return
-        if self.phase_master is not None:
-            self.phase_master.fuse_phases(self.phase_index[p1],self.phase_index[p2])
-        self.phase_index[p2] = self.phase_index[p1]
+        
+        for group_num, group in enumerate(teleporter.get_vertex_groups()):
+            if len(group) == 1: continue
+            self.group_data[group_num] = set(group) # Groups of vertices fused throughout teleportation
+            phase_sum = Fraction(0)
+            for v in group:
+                self.vertex_rank[v] = teleporter.vertex_rank[v]
+                self.vertex_groups[v] = group_num
+                mult = teleporter.phase_mult[v]
+                self.phase_mult[v] = mult # Associated teleportation phase multiplier
+                phase_sum += self.phase(v) * mult
```

Comment:
Is there documentation for how Poly phases works? Or example circuits which have `Poly` phases and should work with phase teleportation?

Reply:
I believe there were some tests somewhere that show how Poly works, but I can't seem to find them now. Many functions for numbers are overloaded to Poly, so you are able to add them together, take the negative, etc. The only thing they miss from the standard FractionLike, is that they don't have denominator and numerator. So adding support could be as simple as just changing the type, and it should Just Work (since phase teleportation only adds together phases or negates them and is agnostic to the actual input).

Reply:
Okay, I already changed the type so that should work. The only time denominator is used is in checking if a vertex is non-Clifford and should be indexed/tracked. Should it automatically treat all Poly phases as non-Clifford for the purposes of phase teleportation?

Reply:
For now that might be simplest.

Reply:
This is done, although i haven't added any Poly tests.